libscidavis/src/Convolution.cpp

```diff
@@ -93,32 +88,31 @@ void Convolution::setDataFromTable(Table *t, const QString &signalColName,
         return;
     }
 
-    d_n = rows;
-
     d_n_signal = 16; // tmp number of points
-    while (unsigned(d_n_signal) < d_n + d_n_response / 2)
+    while (unsigned(d_n_signal) < rows + d_n_response / 2)
         d_n_signal *= 2;
 
-    d_x = new double[d_n_signal]; // signal
-    d_y = new double[d_n_response]; // response
-
-    if (d_y && d_x) {
-        memset(d_x, 0, d_n_signal * sizeof(double)); // zero-pad signal data array
-        for (unsigned i = 0; i < d_n; i++)
-            d_x[i] = d_table->cell(i, signal_col);
-        for (int i = 0; i < d_n_response; i++)
-            d_y[i] = d_table->cell(i, response_col);
-    } else {
```

Comment:
`d_n` is supposed to reflect the size of data created with `new`. This is the only time in the whole project when d_n is assigned arbitrarily, because `d_x` is created with size `d_n_signal` and `d_y` with `d_n_response`. However, iteration for `d_x` goes over `d_n`. Is it bug or I don't understand something?

Reply:
No that is correct. d_n_signal is the power of two that is just greater of equal to the number of rows. d_x is filled with zeros before starting, and then rows data is copied into it.

Reply:
The way I would do it is make d_x and d_y vectors. You can do d_x=move(vector<double>(d_n_signal)). For d_y, better to use clear, reserve and push_back to avoid an unnecessary zero initialisation. You can get rid of d_n_signal and d_n_response as member variables, since they are just the sizes of d_x and d_y.

Reply:
>No that is correct. d_n_signal is the power of two that is just greater of equal to the number of rows. d_x is filled with zeros before starting, and then rows data is copied into it.

No doubt. But the iteration goes over `d_n` which is not necessarily equal to `d_n_signal`. Some rows are left uncopied from `signal_col` to `d_x`.

>The way I would do it is make d_x and d_y vectors. You can do d_x=move(vector(d_n_signal)). For d_y, better to use clear, reserve and push_back to avoid an unnecessary zero initialisation. 

`d_x` and `d_y` are made to be vectors in this PR. Actually, all of this hassle is to make 'em `std::vector`s (see `Filter.h` lines 136-139). Also next lines resize and fill vectors with zeroes. No point using `push_back` when the size is known.

>You can get rid of d_n_signal and d_n_response as member variables, since they are just the sizes of d_x and d_y.

Then I'm to remove `d_n` (or `d_n`) also, because what does it represent then? Size of what?

Reply:
d_n represents the number of rows copied into the data. At the end of the routine, that many items are copied back out.

Think FFT, where the input needs to be a power of 2. If the input isn't, then this algorithm zero pads to the next power of two.

resize() only zeros out elements added to the vector. clear() followed by resize() zeros the entire vector. Actually, strictly speaking, only the last d_n_signal-d_n  need to be zeroed.

Unnecessary zeroing of vectors can be a source of performance drag - which is why reserve() (which allocates the memory, but doesn't zero it), followed by push_back of the data is a good choice. In d_x's case, you can do a resize after the data is copied in, which zeros out the remaining elements.

In any case, SciDAVis is not exactly a high performance application, so it really doesn't matter much. I just like to "do things right" :).

Reply:
Sorry - it appears I didn't fully review this PR. I've run out of time today - I'll swing back on this tomorrow.

Reply:
> d_n represents the number of rows copied into the data. At the end of the routine, that many items are copied back out.
>
> Think FFT, where the input needs to be a power of 2. If the input isn't, then this algorithm zero pads to the next power of two.

You're right, I'll fix it.

> resize() only zeros out elements added to the vector. clear() followed by resize() zeros the entire vector. Actually, strictly speaking, only the last d_n_signal-d_n need to be zeroed.
> 
> Unnecessary zeroing of vectors can be a source of performance drag - which is why reserve() (which allocates the memory, but doesn't zero it), followed by push_back of the data is a good choice. In d_x's case, you can do a resize after the data is copied in, which zeros out the remaining elements.
> 
> In any case, SciDAVis is not exactly a high performance application, so it really doesn't matter much. I just like to "do things right" :).

Right now I'm removing all `d_n()` calls from the code since and replacing with appropriate `std::vector::size()`. Also, I will rewrite the initializers to `resize()` and `push_back()` where possible.